exporter/awsprometheusremotewriteexporter/factory.go

```diff
@@ -0,0 +1,127 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsprometheusremotewriteexporter
+
+import (
+	"context"
+	"errors"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/confighttp"
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+	prw "go.opentelemetry.io/collector/exporter/prometheusremotewriteexporter"
+)
+
+const (
+	typeStr = "awsprometheusremotewrite" // The value of "type" key in configuration.
+)
+
+// NewFactory returns a factory of the AWS Prometheus Remote Write exporter that can be registered to the Collector.
+func NewFactory() component.ExporterFactory {
+	return exporterhelper.NewFactory(
+		typeStr,
+		createDefaultConfig,
+		exporterhelper.WithMetrics(createMetricsExporter))
+}
+
+func createMetricsExporter(_ context.Context, params component.ExporterCreateParams,
+	cfg configmodels.Exporter) (component.MetricsExporter, error) {
+	// check if the configuration is valid
+	prwCfg, ok := cfg.(*Config)
+	if !ok {
+		return nil, errors.New("invalid configuration")
+	}
+
+	if !validateAuthConfig(prwCfg.AuthSettings) {
+		return nil, errors.New("invalid authentication configuration")
+	}
+
+	client, cerr := prwCfg.HTTPClientSettings.ToClient()
+	if cerr != nil {
+		return nil, cerr
+	}
+
+	// load AWS auth configurations and create interceptor based on configuration
+	if applyAuth(prwCfg.AuthSettings) {
```

Comment:
we can change HttpClientConfig to allow "CustomRoundTripper" in core:

```golang
HTTPClientSettings {
	...

	CustomRoundTripper func(next http.RoundTripper) http.RoundTripper
}

func (hcs *HTTPClientSettings) ToClient() (*http.Client, error) {
	...

	if (hcs.CustomRoundTripper != nil {
		clientTransport = hcs.CustomRoundTripper(clientTransport)
	}

	return &http.Client{
		Transport: clientTransport,
		Timeout:   hcs.Timeout,
	}, nil
}
```

Then you can use the same trick to embed the prw.Factory into this factory and override (it is possible) the default config and the CreateMetricsExporter funcs.

Reply:
interesting, will try this out 👍

Reply:
> Then you can use the same trick to embed the prw.Factory into this factory and override (it is possible) the default config and the CreateMetricsExporter funcs.

Could you elaborate on this part of your comment? I'm not quite sure what you mean by embedding the prw.Factory into this factory. 

Do you mean to create a factory using `prw.NewFactory()` within this exporter's `NewFactory()`, and then modify its fields to override the default config and createMetricsExporter function? If so I don't understand the benefit compared to the current approach

Reply:
@bogdandrutu

Reply:
As I understand it, this would streamline the code. It would require an issue in the core repo, a PR to add the custom RoundTripper support in core, a TODO in the code here saying what's going to happen, an issue in the repo here saying this will happen, and lastly a PR to change the code here. If you agree, probably OK to resolve this issue and begin filing new issues.

Reply:
I think i've already made the necessary changes in the core repo: https://github.com/open-telemetry/opentelemetry-collector/pull/2085. But it wasn't documented with issues, my bad for that.

Reply:
@amanbrar1999 here is what embed means, and how the whole factory becomes way simpler:

```golang

type AwsFactory struct {
   component.ExporterFactory
}

func NewFactory() component.ExporterFactory {
	return  &AwsFactory{ExporterFactory : prw.NewFactory()}
}

func (af *AwsFactory) CreateDefaultConfig() configmodels.Exporter {
  cfg := &Config {
    Config: ExporterFactory.CreateDefaultConfig(),
    AuthSettings: AuthSettings{
      Region:  "",
      Service: "",
    },
  }
  
  cfg.HTTPClientSettings.CustomRoundTripper(func(next http.RoundTripper) (http.RoundTripper, error) {
     	if !validateAuthConfig(cfg.AuthSettings) {
		return nil, errors.New("invalid authentication configuration")
	}

	newSigningRoundTripper(cfg.AuthSettings, next)
     })
  return cfg
}

func newSigningRoundTripper(auth AuthSettings, next http.RoundTripper) (http.RoundTripper, error) {
	if !applyAuth(cfg.AuthSettings) {
		return next, nil
	}

	sess, err := session.NewSession(&aws.Config{
		Region: aws.String(cfg.AuthSettings.Region)},
	)
	if err != nil {
		return nil, err
	}

	if _, err = sess.Config.Credentials.Get(); err != nil {
		return nil, err
	}

	// Get Credentials, either from ./aws or from environmental variables
	creds := sess.Config.Credentials
	signer := v4.NewSigner(creds)

	rtp := SigningRoundTripper{
		transport: roundTripper,
		signer:    signer,
		cfg:       sess.Config,
		service:   cfg.AuthSettings.Service,
	}

	// return a RoundTripper
	return &rtp, nil
}

```

Reply:
Ah I see, that makes a lot of sense. Thank you so much! I will make these changes tomorrow as today is a holiday in Canada

Reply:
Also if you want the default config to not inherit default values from the embedded factory just write the entire config from scratch as you currently do, not sure if that is what you want or you want to reuse the default values from the prw (did not check if you change any other value).